### ISSUE TITLE
[Regression] Gradient explodes after upgrading to JAX 0.4.33 from 0.4.30

### ISSUE THREAD

**Issue by qGentry** [jax-ml/jax#24114]
### Description

I'm training LLAMA-3.1-like transformer architecture in Hybrid Sharded Data Parallel-Context Parallel setup on 32GPUs.
Upgrading to JAX 0.4.33 has broken training of 70B model - loss becomes NaN after single training step.
Evidences that I've collected so far:
* Loss on the first step is exactly the same on 0.4.33 and 0.4.30
<img width="1307" alt="Screenshot 2024-10-04 at 11 50 55" src="https://github.com/user-attachments/assets/035dcc1c-96ff-4103-bd17-67eb5e799e66">

* Gradients of the unembedding layer and last layer norm are also exactly the same on the first step.
<img width="541" alt="Screenshot 2024-10-04 at 11 51 57" src="https://github.com/user-attachments/assets/856a0308-a725-4b68-b728-1a2dbc250e0b">

* Gradient already explodes for the last transformer layer's MLP hidden->output layer weight matrix, which I believe is the first layer after token_unembedding and last layer norm matrix.
<img width="1306" alt="Screenshot 2024-10-04 at 11 54 35" src="https://github.com/user-attachments/assets/ff60ea54-de1e-49c9-8527-f0173211981e">

* On JAX 0.4.30, 0.4.29 I've trained tens of such models with different hyperparams and datasets and have never seen any NaN.

* For now, I wasn't able to reproduce this behavior on smaller models, but I'm working on it.

* XLA dumps attached
[xla_dump_0_4_30.tar.gz](https://github.com/user-attachments/files/17256558/xla_dump_0_4_30.tar.gz)
[xla_dump_0_4_33.tar.gz](https://github.com/user-attachments/files/17256560/xla_dump_0_4_33.tar.gz)

### System info (python version, jaxlib version, accelerator, etc.)

```
Python 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import jax; jax.print_environment_info()
jax:    0.4.33
jaxlib: 0.4.33
numpy:  1.24.3
python: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
jax.devices (8 total, 8 local): [CudaDevice(id=0) CudaDevice(id=1) ... CudaDevice(id=6) CudaDevice(id=7)]
process_count: 1
platform: uname_result(system='Linux', node='computeinstance-e00xy41pgq1s49hjc5', release='5.15.0-118-generic', version='#128-Ubuntu SMP Fri Jul 5 09:28:59 UTC 2024', machine='x86_64')


$ nvidia-smi
Fri Oct  4 10:07:59 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.90.07              Driver Version: 550.90.07      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA H100 80GB HBM3          On  |   00000000:8D:00.0 Off |                    0 |
| N/A   28C    P0            110W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA H100 80GB HBM3          On  |   00000000:91:00.0 Off |                    0 |
| N/A   27C    P0            110W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA H100 80GB HBM3          On  |   00000000:95:00.0 Off |                    0 |
| N/A   30C    P0            110W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA H100 80GB HBM3          On  |   00000000:99:00.0 Off |                    0 |
| N/A   27C    P0            112W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   4  NVIDIA H100 80GB HBM3          On  |   00000000:AB:00.0 Off |                    0 |
| N/A   28C    P0            109W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   5  NVIDIA H100 80GB HBM3          On  |   00000000:AF:00.0 Off |                    0 |
| N/A   26C    P0            109W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   6  NVIDIA H100 80GB HBM3          On  |   00000000:B3:00.0 Off |                    0 |
| N/A   29C    P0            112W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   7  NVIDIA H100 80GB HBM3          On  |   00000000:B7:00.0 Off |                    0 |
| N/A   27C    P0            110W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
+-----------------------------------------------------------------------------------------+
```

[XLA issue](https://github.com/openxla/xla/issues/17922)

**Comment by qGentry** [comment]
Hi!
Bisecting Jax version would be quite time-consuming for me as I would have to rebuild my image for every run.
I've also tried freshly release JAX 0.4.34 - problem hasn't been solved there.
Please also check xla issue, I've added tons of information and evidences there. Specifically, I've bisected specific model size at which something clearly breaks.
https://github.com/openxla/xla/issues/17922#issuecomment-2394209837

**Comment by qGentry** [comment]
We've figured our workaround for this problem - https://github.com/openxla/xla/issues/17922#issuecomment-2396280557

**Comment by qGentry** [comment]
PR with this fix will be merged to XLA soon
